[2022-03-na-group1] Implement BucketFill

antoine-agthe-unity opened the pull request:
This PR implements the system, components and helpers types for the bucket fill state.

I've separated static methods into Helper types.
Tell me if it makes sense for you.

antoine-agthe-unity committed:
Adds the BucketFill component

antoine-agthe-unity committed:
Adds the BucketFill component to the Fetcher and the Omniworker

antoine-agthe-unity committed:
Adds the BucketHelper class

Provides static methods to manipulate the components of a bucket

antoine-agthe-unity committed:
Adds the WaterPool Helper class

Provides static methods to manipulate the components of a water pool

antoine-agthe-unity committed:
Adds the BucketFill system

antoine-agthe-unity committed:
Merge branch 'dots-training-2022-03-na-group1' into dots-training-2022-03-na-group1-/bucketFillSystem

antoine-agthe-unity committed:
Removes BucketFill from the Components namespace

UnityPix commented on Ported/dots-training-2022-03-na-group1_BucketBrigade/Assets/Scripts/Components/BucketFill.cs:
```diff
@@ -0,0 +1,12 @@
+    /// <summary>
+    ///     The time the filling process is finished.
+    /// </summary>
+    public float cooldown;
```
been following the practice of single value components having Value as their field.

antoine-agthe-unity replied:
So I propose to rename the `BucketFill` component to `Cooldown`, and its single field `Value` according to your comment. Agreed?

---
That being said, and even if it works properly in the scope of our small project, I don't get the point of such a practice.
It looks like a bad choice regarding how we usually iterate on game code.
I might be wrong, so maybe @BrianWill-Unity you could also give your opinion on it.

### Concurrent use

Only one of our system requires a cooldown at the moment, so that is fine to have a `Cooldown` component.
But if another system requires a cooldown aspect, we might be tempted to reuse the `Cooldown` component.
But while a entity can have only one component of a given type, the two systems might need to read/write their own cooldown value at the same moment, which requires two different memory space to store it.
Then `Cooldown` becomes too much general and requires annoying changes described below.

### System requirement can evolve.

At the moment, `BucketFillSystem` only requires a single `cooldown` field. Naming it `Value` is not a short term problem.
But nothing guarantees that the requirements won't evolve and need an additional field. Let's say a flag.
Because of this flag, we would need to rename the component from `Cooldown` back to `BucketFill`, `Value` back to `cooldown` in order to differentiate it from the flag field, and propagate this change across the whole source code, which is not versioning-friendly as it applies easily-avoidable modifications.  
Storing the flag in another `Flag` component is still possible, but it could lead to the "Concurrent use" problem mentioned above.  
Moreover, I'm quite sure that there is still a benefit in grouping values together when they are used together instead of separating them into singled field-ed component for no reason.

antoine-agthe-unity replied:
done

UnityPix commented on Ported/dots-training-2022-03-na-group1_BucketBrigade/Assets/Scripts/Systems/BucketFillSystem.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    protected override void OnUpdate()
+    {
+        var currentTime = UnityEngine.Time.realtimeSinceStartup;
```
Time.elapsed is the version that's modeled in the tutorial.

antoine-agthe-unity replied:
My bad, I was already told about it.
Changing.

antoine-agthe-unity replied:
done

UnityPix commented on Ported/dots-training-2022-03-na-group1_BucketBrigade/Assets/Scripts/Systems/BucketFillSystem.cs:
```diff
@@ -0,0 +1,32 @@
+                if (workerState.Value != WorkerState.FillingBucket) return;
+
+                // Still waiting for the filling to be completed
+                if (fillStatus.cooldown < currentTime) return;
```
if cooldown is when the cooldown is scheduled to end, what sets this?

antoine-agthe-unity replied:
A system doesn't have any "OnEnter" because it is not a state-machine piece of framework.
So only entities properly prepared to be filtering are processed.
Which means that the system responsible of pushing the entity in `WorkerState.FillingBucket` is also responsible of giving it a proper `cooldown` value.

UnityPix replied:
okay.  maybe if you want to add that to the idle code once i merge it.

antoine-agthe-unity committed:
Using the world elapsed time instead of UnityEngine realtimeSinceStartup

antoine-agthe-unity committed:
Rename `BucketFill` to `Cooldown`

Also renaming the `cooldown` field to `Value` to propagate the change across the code.
